sync15-adapter/src/sync_stateful.rs

```diff
@@ -0,0 +1,126 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// This helps you perform a "stateful" sync - ie, a sync which persists state
+// between syncs and uses this state to automatically do the right thing
+// with the stores etc.
+
+use std::result;
+use std::cell::Cell;
+use failure;
+use client::{Sync15StorageClient, Sync15StorageClientInit};
+use state::{GlobalState, SetupStateMachine};
+use sync::{self, Store};
+use key_bundle::KeyBundle;
+use error::Error;
+
+// Info stored in memory about the client to use. We reuse the client unless
+// we discover the client_init has changed, in which case we re-create one.
+pub struct ClientInfo {
+    // the client_init used to create the client.
+    client_init: Sync15StorageClientInit,
+    // the client we will reuse if possible.
+    client: Sync15StorageClient,
+}
+
+// This is used only to persist a GlobalState, which is serializable.
+
+// XXX - should we define this simply as a Cell<>, like last_client_info,
+// and leave the persisting to whoever provides the cell?
+
+// XXX - alternatively/additionally, instead of defining this in terms of
+// the GlobalState structure, we could consider defining it in terms of
+// Serializable/Deserializable?
+pub trait GlobalStateProvider {
+    fn load(&self) -> result::Result<Option<GlobalState>, failure::Error>;
+
+    fn save(&self, state: Option<&GlobalState>) -> result::Result<(), failure::Error>;
+}
+
+pub fn sync_stateful(
+    store: &Store,
```

Comment:
Nit: `&dyn Store` (or `&[&dyn Store]` if we take a slice). IIUC one of the upcoming rust versions deprecates the non-`dyn` syntax for trait objects, and another takes it out entirely. IMO it's better to be clear about it.

Also `&dyn GlobalStateProvider`, since it's also a trait object.